src/models/common_decoder.h

```diff
@@ -227,9 +312,19 @@ class CommonDecoder : public AbstractDecoder {
         int *positionIds = this->getPositionIds(ids, batchSize, inputSeqLen, step + this->prefixSharing);
         t1.release();
 
+        // if current pipeline parallel stage rank isn't the first stage, should receive previous stage data
+        if (ctx->ppSize > 1 && ctx->ppRank > 0) {
+            // [MPI] Recv data from world_rank 0
+            int curr_world_rank = ctx->ppRank * ctx->tpSize + ctx->tpRank;
+            int prev_world_rank = (ctx->ppRank - 1) * ctx->tpSize + ctx->tpRank;
+            MPI_Recv(embBuf, batchSize * inputSeqLen * ctx->hiddenSize, MPI_FLOAT, prev_world_rank, curr_world_rank,
+                    MPI_COMM_WORLD, MPI_STATUS_IGNORE);
+        }
+
```

Comment:
This will reintroduce the MPI dependency into xft.so. It should be included in comm_helper.so and referenced through messager.

Reply:
Error: different scope when dynamic loading so file